lib/spack/spack/fetch_strategy.py

```diff
@@ -337,25 +344,44 @@ def expand(self):
         with working_dir(tarball_container):
             decompress(self.archive_file)
 
-        # Check for an exploding tarball, i.e. one that doesn't expand
-        # to a single directory.  If the tarball *didn't* explode,
-        # move contents up & remove the container directory.
+        if self.stage.expanded:
+            tty.debug('Source already staged to %s' % self.stage.source_path)
+            return
+
+        # Check for an exploding tarball, i.e. one that doesn't expand to
+        # a single directory.  If the tarball *didn't* explode, move its
+        # contents to the staging source directory & remove the container
+        # directory.  If the tarball did explode, just rename the tarball
+        # directory to the staging source directory.
         #
-        # NOTE: The tar program on Mac OS X will encode HFS metadata
-        # in hidden files, which can end up *alongside* a single
-        # top-level directory.  We ignore hidden files to accomodate
-        # these "semi-exploding" tarballs.
+        # NOTE: The tar program on Mac OS X will encode HFS metadata in
+        # hidden files, which can end up *alongside* a single top-level
+        # directory.  We initially ignore presence of hidden files to
+        # accomodate these "semi-exploding" tarballs but ensure the files
+        # are copied to the source directory.
         files = os.listdir(tarball_container)
         non_hidden = [f for f in files if not f.startswith('.')]
         if len(non_hidden) == 1:
-            expanded_dir = os.path.join(tarball_container, non_hidden[0])
-            if os.path.isdir(expanded_dir):
-                for f in files:
-                    shutil.move(os.path.join(tarball_container, f),
-                                os.path.join(self.stage.path, f))
+            # TODO: Is it correct to assume the hidden files should be moved
+            # TODO: to the root of the stage source directory?
```

Comment:
yep they should stay alongside the expanded directory.  Note that this is probably something you need to account for in these recurring lines:

```
        stage_entries = os.listdir(self.stage.path)
        assert len(stage_entries) == 1
        repo_root = os.path.join(self.stage.path, stage_entries[0])
```

So,  in other words, shouldn't he logic here, to check for a single non-hidden directory, be used at the root of the stage to account for the case where you've already moved these files up?

I would make this into a function and use the same logic in the `tarball_container` and in `self.stage.path`.  I think it'll reduce the complexity for this special case too.

Reply:
I'm not quite following the issue here.  

The non-hidden file checks do not apply if the source was already staged.  When they do apply, it is to the tarball_container, os.path.join(self.stage.path, "spack-expanded-archive"), after the archive file is decompressed into it.

Actually, why is the self.stage.expanded check deferred to after the tarball is decompressed?  It seems more efficient to skip all the decompress-related code in this case.

Reply:
I refactored the above function per the initial comment; however, its use is no longer relevant/critical to ensure only one stage directory entry (for moving "pre-expanded" git, hg, or svn fetch strategy content due to changing their strategies to use their output directory options.